### ISSUE TITLE
Solr instance is not running in SolrCloud mode

### ISSUE THREAD

**Issue by ultimike** [ddev/ddev-contrib#221]
I'm attempting to use Drupal 10.0.3 with Search API 4.2.10.

I followed the instructions on the README, but after I set the number of shards to 3 and submit (while trying to upload a new config set from /admin/config/search/search-api/server/ddev_server I get the following error:

```
Configset upload failed with error code 400: Solr HTTP error: OK (400) { "responseHeader":{ "status":400, "QTime":6}, "error":{ "metadata":[ "error-class","org.apache.solr.common.SolrException", "root-error-class","org.apache.solr.common.SolrException"], "msg":"Solr instance is not running in SolrCloud mode.", "code":400}}
```

I am able to access the solr server at http://moreddev.ddev.site:8983/solr/#/

On the Drupal server view page (/admin/config/search/search-api/server/ddev_server):

*  Server Connection: `The Solr server could be reached.`  
*  Collection Connection: `The Solr collection could not be accessed. Further data is therefore unavailable.`  
*  Configured Solr Version: `8.11.2`  
*  Additional information: `An error occurred while trying to retrieve additional information from the Solr server: Solr endpoint http://ddev-moreddev-solr:8983/ not found (code: 404, body: <html> <head> <meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/> <title>Error 404 Not Found</title> </head> <body><h2>HTTP ERROR 404 Not Found</h2> <table> <tr><th>URI:</th><td>/solr/techproducts/admin/luke</td></tr> <tr><th>STATUS:</th><td>404</td></tr> <tr><th>MESSAGE:</th><td>Not Found</td></tr> <tr><th>SERVLET:</th><td>default</td></tr> </table> </body> </html> , message: Solr HTTP error: OK (404) <html> <head> <meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/> <title>Error 404 Not Found</title> </head> <body><h2>HTTP ERROR 404 Not Found</h2> <table> <tr><th>URI:</th><td>/solr/techproducts/admin/luke</td></tr> <tr><th>STATUS:</th><td>404</td></tr> <tr><th>MESSAGE:</th><td>Not Found</td></tr> <tr><th>SERVLET:</th><td>default</td></tr> </table> </body> </html> ).`

What am I missing?

thanks,
-mike

**Comment by rfay** [comment]
Hi @ultimike - Unfortunately I haven't ever been able to make this solr cloud version work, and https://github.com/ddev/ddev-contrib/pull/195 never got far enough. @mkalkbrenner may be able to help here, but he's pretty busy.

I'm afraid, much to @mkalkbrenner's disapproval, I have to recommend the maintained, tested and stable traditional solr installation technique, `ddev get ddev/ddev-drupal9-solr`, https://github.com/ddev/ddev-drupal9-solr

**Comment by gitressa** [comment]
I created #227.

**Comment by rfay** [comment]
I'd sure rather have you create a solr cloud add-on that works! And then maintain it. Would you consider that?

**Comment by gitressa** [comment]
Thanks for the confidence in me @rfay! But I am afraid adjusting existing configuration is as far as my very limited Solr/Zookeeper/Docker abilities go ...

I would much prefer to get you, @ultimike and the original author @mkalkbrenner to review the PR, if at all possible.

In the meantime, perhaps you can check if it works? On my local DDEV, both standalone Solr and Solr Cloud work well, using #227. It would be great to get this confirmed from multiple sources.

**Comment by rfay** [comment]
Following your d.o issue, and maybe we can get @mkalkbrenner (maintainer of search_api_solr) to speak up. And of course if you haven't carefully read and tried https://github.com/ddev/ddev-contrib/pull/195 please take a look at that.

**Comment by rfay** [comment]
The bottom line for me is that the classic recipe is supported and lots of people use it. I've never been successful with the newer multi-instance setup, and don't know how it adds value in a local dev setup, but I'm very naive about solr.  https://github.com/ddev/ddev-drupal9-solr is supported and recommended, and this multi-solr is not supported by anybody I know of, although I keep hoping to get @mkalkbrenner supporting it, but he's a busy person.

**Comment by gitressa** [comment]
Yes, I agree it would be great if the maintainer of Solr @mkalkbrenner at some point had time to add  the steps required to get https://github.com/docker-solr/docker-solr-examples/blob/master/docker-compose/docker-compose.yml working, so that after setting everything up with the excellent documentation on [Apache Solr (Cloud) Integration for DDEV-Local](https://github.com/ddev/ddev-contrib/blob/master/docker-compose-services/solr/README.md) it is possible for Docker/Solr/Zookeeper non-experts to deploy a Drupal 10 Solr Cloud to production, and push it over the goal line.

I managed to get https://github.com/docker-solr/docker-solr-examples/blob/master/docker-compose/docker-compose.yml up and running. But I know too little about Docker/Solr/Zookeeper to add the missing pieces in the puzzle. It may just be adjusting a few lines in _docker-solr-examples/blob/master/docker-compose/docker-compose.yml_ to integrate _solr/security.json_, which is missing?

Like you describe, I used the classic single core Solr instead, and it works perfectly, except for the known issue #15. I get great performance. Solr Cloud may be advantageous, if you have heavy loads, and hundreds of simultaneous users? But many web sites don't have that need, so Solr Cloud would be overkill.

I think that after installing Solr with `ddev get ddev/ddev-drupal9-solr` that the start up slows down, so I have created #228.

**Comment by mkalkbrenner** [comment]
Strange, I run multiple projects with ddev and Solr Cloud. The name is always `ddev-<project>-solr1:8983`, even in these I setup recently.

**Comment by rfay** [comment]
If you're using the https://github.com/ddev/ddev-contrib/blob/master/docker-compose-services/solr/docker-compose.solr.yaml#L20-L23 then the container names will definitely be `solr1`, `solr2`, etc.

**Comment by mkalkbrenner** [comment]
> If you're using the https://github.com/ddev/ddev-contrib/blob/master/docker-compose-services/solr/docker-compose.solr.yaml#L20-L23 then the container names will definitely be solr1, solr2, etc.

Ok, you're right. And I think I now understand what is confusing everyone. I recommend to use
https://github.com/ddev/ddev-contrib/blob/master/docker-compose-services/solr/docker-compose.solr-standalone.yaml

This is sufficient for local development. It isn't the classic standalone mode you might have in mind. It a Solr Cloud with just one node.

> and don't know how it adds value in a local dev setup

Beside the fact that it is much easier to perform Solr development, there're some APIs and features that aren't available if you choose the classic standalone mode:

- https://solr.apache.org/guide/8_11/streaming-expressions.html
- https://solr.apache.org/guide/8_11/collections-api.html
- https://solr.apache.org/guide/8_11/configsets-api.html
- ...

But they're available in the "standalone" the ddev contrib module provides.

The setup with 3 nodes is just required if you really need to simulate or test cluster situations where you have shards, replication, worker nodes etc.

So instead changing the README and rename solr to solr1 we should change emphasize the "standalone cloud mode".

**Comment by rfay** [comment]
OK. The "main" docker-compose.solr.yaml here in ddev-contrib one would assume to be the "normal" one. But we should just remove that one and make the standalone one the main one, right? 

Hopefully we can get an add-on out of this yet :)

**Comment by mkalkbrenner** [comment]
I just released new versions of solarium and search_api_solr this week.
I'll read some documentation and hopefully find some time next week to convert this ddev-contrib into a plugin.
Afterwards we can see how the documentation should be adjusted.

**Comment by rfay** [comment]
Awesome!

**Comment by gitressa** [comment]
Sounds fantastic @mkalkbrenner, thanks! You do an amazing job maintaining the module. It's a lot of work, with many moving parts, and I very much appreciate your efforts.

**Comment by mkalkbrenner** [comment]
@rfay 
https://github.com/mkalkbrenner/ddev-solr

**Comment by mkalkbrenner** [comment]
I'll change the documentation within search_api_solr to use that add-on.
But I'll wait a bit for the decision regarding https://github.com/ddev/ddev/issues/5002

**Comment by rfay** [comment]
Awesome!
